File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        last_lower = [-1] * 26
        first_upper = [-1] * 26
        
        for i, char in enumerate(word):
            if char.islower():
                last_lower[ord(char) - 97] = i
            else:
                idx = ord(char) - 65
                if first_upper[idx] == -1:
                    first_upper[idx] = i
                    
        ans = 0
        for i in range(26):
            if last_lower[i] != -1 and first_upper[i] != -1 and last_lower[i] < first_upper[i]:
                ans += 1
                
        return ans

File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.java
class Solution {
    public int numberOfSpecialChars(String word) {
       int[] lastLower = new int[26];
        int[] firstUpper = new int[26];
        
        for (int i = 0; i < 26; i++) {
            lastLower[i] = -1;
            firstUpper[i] = -1;
        }
        
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (c >= 'a' && c <= 'z') {
                lastLower[c - 'a'] = i;
            } else {
                int idx = c - 'A';
                if (firstUpper[idx] == -1) {
                    firstUpper[idx] = i;
                }
            }
        }
        
        int ans = 0;
        for (int i = 0; i < 26; i++) {
            if (lastLower[i] != -1 && firstUpper[i] != -1 && lastLower[i] < firstUpper[i]) {
                ans++;
            }
        }
        
        return ans; 
    }
}